### PR TITLE
Fix magic number check

### PR DESCRIPTION
**Pull request by rock3r**
This PR fixes a number of issues in the `MagicNumber` check:
 * flagging non-numeric constant expressions (e.g., `true`, `null`)
 * inability to ignore floating point values correctly (due to casting to `long`)
 * incorrect hex number parsing in some cases
 * silently ignores invalid `ignoredNumbers` configuration parameters
 * ignores negative values when the corresponding positive number is
   ignored (e.g., -2 is ignored when 2 is in `ignoredNumbers`)
 * floating point numbers are not correctly parsed in some cases
 * numeric literals that contain underscores are not correctly parsed

I noticed these issues because I got false positives all over the place in [my codebase](https://squanchy.net) as soon as I updated. Unfortunately the tests for the rule in this case were omitting any false positive and any edge case, so that the rule seemed formally valid but was actually severely bugged. I would recommend double checking the test suite for new rules before shipping them in the future.

To be fair, the test suite is still not optimal, but at least captures all the bugs I identified so far. There can/will be more bugs, possibly, but this seems a good starting point. It's not Checkstyle's omonimous check, and the tests are not nearly as exhaustive, but better than it was.

On top of that there are a couple of minor tweaks to make the testing of the class easier (not that Spek ever worked for me in the IDE... but at least from the command line it does), removes an unused Gradle property, adds missing spaces in the `MagicNumber` check description, and renames the `ignoreNumbers` config setting to `ignoredNumbers`.

The rule still doesn't parse binary numbers because it was originally not, but that could be a separate PR if required.

### CODE REVIEW

File: detekt-api/src/main/kotlin/io/gitlab/arturbosch/detekt/api/Issue.kt [hunk 1]
```diff
@@ -11,6 +11,10 @@ data class Issue(val id: String,
 	init {
 		validateIdentifier(id)
 	}
+
+	override fun toString(): String {
+		return "Issue(id='$id', severity=$severity, debt=$debt)"
```
Comment (rock3r) [inline detekt-api/src/main/kotlin/io/gitlab/arturbosch/detekt/api/Issue.kt]: I purposely omitted the `description` here as it is very noisy to have around and it can be univocally be determined based on the `id` anyway

File: detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/MagicNumber.kt [hunk 2]
```diff
@@ -7,47 +7,86 @@ import io.gitlab.arturbosch.detekt.api.Entity
...
+					"By default, -1, 0, 1, and 2 are not considered to be magic numbers.", Debt.TEN_MINS)
 
-	private val ignoreNumbers = valueOrDefault(IGNORE_NUMBERS, "-1,0,1,2")
+	private val ignoredNumbers = valueOrDefault(IGNORED_NUMBERS, "-1,0,1,2")
 			.split(",")
-			.map { it.toLongOrNull() }
-			.filterNotNull()
+			.filterNot { it.isEmpty() }
+			.map { parseAsDouble(it) }
+			.sorted()
```
Comment (rock3r) [inline detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/MagicNumber.kt]: Given this is looked up with `contains`, making sure the contents are sorted should allow for faster searching, were this to be using binary search or something.

**Comment by arturbosch** [comment]
Thank you for your work! I've also found a false positive: int and double etc properties shouldn't be flagged as magic number as they have a meaningful name.

And yes your right we need to double check new rules. As a bunch of new rules get contributed every month I can't check them all out locally, I do not have the time for it. My proposal is to let new rules PR open for at least 2 days so others can also have a look at them and let new rules be deactivate for two version.

**Comment by rock3r** [comment]
That sounds like a reasonable policy. I know you already run Detekt against itself when cutting releases, but I think doing it against a couple more projects when adding new rules,  and quickly glance at the results, could be useful. I noticed the issues immediately after upgrading when running against my side project, and I generally keep up to date with Detekt reasonably closely because this sort of issues is sort of expected in such a young static analysis tool, and I hope for fixes :)
